File: testpy/aes.py
import json
import os
import secrets
import hashlib
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend


class AES_Cryptor:
    def __init__(self, name, key=None, key_seed=None, key_length=32):
        self.name = name
        self.key = key
        self.key_seed = key_seed
        self.key_length = key_length

    def __name__(self):
        return self.name

    # ------------------------------------------------------------
    # Tạo seed số ngẫu nhiên (6 số)
    # ------------------------------------------------------------
    def gen_seed(self):
        """
        Sinh 6 số ngẫu nhiên trong [0,1000), dùng để tạo key deterministically.
        """
        key_seed = [secrets.randbelow(1000) for _ in range(6)]
        self.key_seed = "-".join(str(x).zfill(3) for x in key_seed)
        return self.key_seed

    # ------------------------------------------------------------
    # Tạo key AES-256 từ seed số, dùng HKDF-SHA256 (deterministic)
    # ------------------------------------------------------------
    def gen_key(self):
        """
        Generate secret key từ list seed_numbers (deterministic, dùng HKDF-SHA256).
        """
        if not self.key_seed:
            self.gen_seed()

        # Ghép seed thành chuỗi ví dụ: "123,456,789,321,654,987"
        seed_bytes = self.key_seed.encode("utf-8")

        # Dùng SHA256 để nén lại làm input cho HKDF (optional nhưng tốt)
        seed_digest = hashlib.sha256(seed_bytes).digest()

        # Dùng HKDF để derive key (deterministic, bảo mật)
        hkdf = HKDF(
            algorithm=hashes.SHA256(),
            length=self.key_length,
            salt=b"fixed-hkdf-salt-v1",  # có thể đổi tùy app version
            info=b"aes-key-derivation",
        )

        self.key = hkdf.derive(seed_digest)
        return self.key

    # ------------------------------------------------------------
    # AES-GCM Encrypt / Decrypt
    # ------------------------------------------------------------
    def encrypt_data(self, plaintext, nonce=None):
        """
        Mã hóa dữ liệu với AES-256-GCM.
        - Trả về: (ciphertext, nonce, tag)
        """
        if not self.key:
            raise ValueError("Key chưa được tạo. Hãy gọi gen_key() trước.")

        if not nonce:
            nonce = os.urandom(12)  # Nonce unique
        cipher = Cipher(algorithms.AES(self.key), modes.GCM(nonce), backend=default_backend())
        encryptor = cipher.encryptor()
        ciphertext = encryptor.update(plaintext.encode("utf-8")) + encryptor.finalize()
        tag = encryptor.tag
        return ciphertext, nonce, tag

    def decrypt_data(self, key, ciphertext, nonce, tag):
        """
        Giải mã dữ liệu với AES-256-GCM.
        """
        cipher = Cipher(algorithms.AES(key), modes.GCM(nonce, tag), backend=default_backend())
        decryptor = cipher.decryptor()
        plaintext = decryptor.update(ciphertext) + decryptor.finalize()
        return plaintext

# ------------------------------------------------------------
    # SAVE / LOAD KEY
    # ------------------------------------------------------------
    def save_key(self, file_path, save_seed=True):
        """
        Lưu key xuống file JSON.
        - Nếu save_seed=True: chỉ lưu seed (bảo mật hơn, có thể tái sinh key)
        - Nếu save_seed=False: lưu luôn full key bytes (ít bảo mật hơn)
        """
        if save_seed:
            if not self.key_seed:
                raise ValueError("Chưa có seed để lưu. Gọi gen_seed() trước.")
            data = {
                "type": "seed",
                "name": self.name,
                "key_seed": self.key_seed,
                "checksum": hashlib.sha256(",".join(map(str, self.key_seed)).encode()).hexdigest()[:16],
            }
        else:
            if not self.key:
                raise ValueError("Chưa có key để lưu. Gọi gen_key() trước.")
            data = {
                "type": "key",
                "name": self.name,
                "key_hex": self.key.hex(),
                "checksum": hashlib.sha256(self.key).hexdigest()[:16],
            }

        # os.makedirs(os.path.dirname(file_path), exist_ok=True)
        with open(file_path, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=4)
        print(f"[{self.name}] Saved {data['type']} to {file_path}")

    def load_key(self, file_path):
        """
        Load seed hoặc key từ file JSON.
        - Nếu type='seed' → tái tạo key bằng HKDF.
        - Nếu type='key' → dùng key trực tiếp.
        """
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"Không tìm thấy file: {file_path}")

        with open(file_path, "r", encoding="utf-8") as f:
            data = json.load(f)

        if data["type"] == "seed":
            self.key_seed = data["key_seed"]
            self.gen_key()
            print(f"[{self.name}] Loaded seed & regenerated key.")
        elif data["type"] == "key":
            self.key = bytes.fromhex(data["key_hex"])
            print(f"[{self.name}] Loaded raw key from file.")
        else:
            raise ValueError("File không hợp lệ (thiếu type=seed|key).")
        return self.key
    
def NewAES(name, key=None, key_seed=None, key_length=32):
    return AES_Cryptor(name, key, key_seed, key_length)

File: internal/crypto/aes/aes.go
package aes

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"os"
	"path/filepath"
	"strings"

	"golang.org/x/crypto/hkdf"
)

type AESCryptor struct {
	Name      string
	Key       []byte
	KeySeed   string
	KeyLength int
}

func NewAES(name string, key []byte, keySeed string, keyLength int) *AESCryptor {
	if keyLength == 0 {
		keyLength = 32
	}
	return &AESCryptor{Name: name, Key: key, KeySeed: keySeed, KeyLength: keyLength}
}

func (a *AESCryptor) GetName() string {
	return a.Name
}

// GenSeed generates a random seed string consisting of 6 numbers (0-999) joined by "-".
func (a *AESCryptor) GenSeed() string {
	var seeds []string
	for range 6 {
		r := randBelow(1000)
		seeds = append(seeds, fmt.Sprintf("%03d", r))
	}
	a.KeySeed = strings.Join(seeds, "-")
	return a.KeySeed
}

// GenKey generates the AES key from the seed using HKDF-SHA256.
func (a *AESCryptor) GenKey() ([]byte, error) {
	if a.KeySeed == "" {
		a.GenSeed()
	}
	seedBytes := []byte(a.KeySeed)
	seedDigest := sha256.Sum256(seedBytes)
	h := hkdf.New(sha256.New, seedDigest[:], []byte("fixed-hkdf-salt-v1"), []byte("aes-key-derivation"))
	key := make([]byte, a.KeyLength)
	_, err := io.ReadFull(h, key)
	if err != nil {
		return nil, err // To match Python's raising behavior
	}
	a.Key = key
	return key, nil
}

// EncryptData encrypts the plaintext using AES-256-GCM.
// Returns ciphertext, nonce, tag separately.
func (a *AESCryptor) EncryptData(plaintext string, nonce []byte) ([]byte, []byte, []byte, error) {
	return a.EncryptBytes([]byte(plaintext), nonce)
}

func (a *AESCryptor) EncryptBytes(data []byte, nonce []byte) ([]byte, []byte, []byte, error) {

	if len(a.Key) == 0 {
		return nil, nil, nil, fmt.Errorf("key not generated")
	}

	if nonce == nil {
		nonce = make([]byte, 12)
		rand.Read(nonce)
	}

	block, err := aes.NewCipher(a.Key)
	if err != nil {
		return nil, nil, nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, nil, nil, err
	}

	out := gcm.Seal(nil, nonce, data, nil)

	ciphertext := out[:len(out)-16]
	tag := out[len(out)-16:]

	return ciphertext, nonce, tag, nil
}

// DecryptData decrypts the ciphertext using AES-256-GCM.
// This is compatible with the Python version and requires the key separately.
func (a *AESCryptor) DecryptData(key, ciphertext, nonce, tag []byte) (string, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	ctWithTag := append(ciphertext, tag...)
	plain, err := gcm.Open(nil, nonce, ctWithTag, nil)
	if err != nil {
		return "", err
	}
	return string(plain), nil
}

// SaveKey saves the key or seed to a JSON file.
func (a *AESCryptor) SaveKey(filePath string) {
	var data map[string]string

	if a.KeySeed == "" {
		panic("No seed to save. Call GenSeed() first.")
	}
	// Mimic the Python code exactly, including the character-wise join (potential original bug, but for compatibility)
	joined := strings.Join(strings.Split(a.KeySeed, ""), ",")
	sum := sha256.Sum256([]byte(joined))
	checksum := hex.EncodeToString(sum[:])[:16]
	data = map[string]string{
		"type":     "seed",
		"name":     a.Name,
		"key_seed": a.KeySeed,
		"checksum": checksum,
	}

	jsonData, err := json.MarshalIndent(data, "", "    ")
	if err != nil {
		panic(err)
	}
	err = os.MkdirAll(filepath.Dir(filePath), 0755)
	if err != nil {
		panic(err)
	}
	err = os.WriteFile(filePath, jsonData, 0644)
	if err != nil {
		panic(err)
	}
	fmt.Printf("[%s] Saved %s to %s\n", a.Name, data["type"], filePath)
}

// LoadKey loads the key or seed from a JSON file and sets it on the instance.
func (a *AESCryptor) LoadKey(filePath string) []byte {
	buf, err := os.ReadFile(filePath)
	if err != nil {
		panic(err)
	}
	var data map[string]string
	err = json.Unmarshal(buf, &data)
	if err != nil {
		panic(err)
	}
	typ, ok := data["type"]
	if !ok {
		panic("File invalid (missing type=seed|key).")
	}
	switch typ {
	case "seed":
		seed, ok := data["key_seed"]
		if !ok {
			panic("Missing key_seed in file.")
		}
		a.KeySeed = seed
		a.GenKey()
		fmt.Printf("[%s] Loaded seed & regenerated key.\n", a.Name)
	case "key":
		kh, ok := data["key_hex"]
		if !ok {
			panic("Missing key_hex in file.")
		}
		a.Key, err = hex.DecodeString(kh)
		if err != nil {
			panic(err)
		}
		fmt.Printf("[%s] Loaded raw key from file.\n", a.Name)
	default:
		panic("File invalid (type must be seed|key).")
	}
	return a.Key
}

func randBelow(n int) int {
	bi, err := rand.Int(rand.Reader, big.NewInt(int64(n)))
	if err != nil {
		panic(err)
	}
	return int(bi.Int64())
}
